packages/libs/deer/json/src/lib.rs

```diff
@@ -0,0 +1,499 @@
+            Some(value) => Err(Report::new(TypeError.into_error())
+                .attach(ExpectedType::new(<$reflection as Reflection>::document()))
+                .attach(ReceivedType::new(into_document(&value))))
+                .change_context(DeserializerError),
+            None => Err(Report::new(MissingError.into_error())
+                .attach(ExpectedType::new(<$reflection as Reflection>::document()))
+                .change_context(DeserializerError)),
+        }
+    };
+}
+
+struct Deserializer {
+    value: Option<Value>,
+}
+
+impl From<Value> for Deserializer {
+    fn from(value: Value) -> Self {
+        Self { value: Some(value) }
+    }
+}
+
+impl Deserializer {
+    const fn empty() -> Self {
+        Self { value: None }
+    }
+}
+
+impl<'de> deer::Deserializer<'de> for Deserializer {
+    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, DeserializerError>
+    where
+        V: Visitor<'de>,
+    {
+        match self.value {
+            None => visitor.visit_none(),
+            Some(Value::String(string)) => visitor.visit_string(string),
+            Some(Value::Null) => visitor.visit_null(),
+            Some(Value::Bool(bool)) => visitor.visit_bool(bool),
+            Some(Value::Number(number)) => serde_to_deer_number(&number)
+                .ok_or_else(|| todo!())
+                .and_then(|number| visitor.visit_number(number)),
+            Some(Value::Array(array)) => visitor.visit_array(ArrayAccess::new(array)),
+            Some(Value::Object(object)) => visitor.visit_object(ObjectAccess::new(object)),
+        }
+        .change_context(DeserializerError)
+    }
+
+    fn deserialize_none<V>(self, visitor: V) -> Result<V::Value, DeserializerError>
+    where
+        V: Visitor<'de>,
+    {
+        self.value.map_or_else(
+            || visitor.visit_none().change_context(DeserializerError),
+            |value| {
+                Err(Report::new(TypeError.into_error())
+                    .attach(ExpectedType::new(NoneReflection::document()))
+                    .attach(ReceivedType::new(into_document(&value)))
+                    .change_context(DeserializerError))
+            },
+        )
+    }
+
+    fn deserialize_null<V>(self, visitor: V) -> Result<V::Value, DeserializerError>
+    where
+        V: Visitor<'de>,
+    {
+        match self.value {
+            Some(Value::Null) => visitor.visit_null().change_context(DeserializerError),
+            Some(value) => Err(Report::new(TypeError.into_error())
+                .attach(ExpectedType::new(NullReflection::document()))
+                .attach(ReceivedType::new(into_document(&value)))
+                .change_context(DeserializerError)),
+            None => Err(Report::new(MissingError.into_error())
+                .attach(ExpectedType::new(NullReflection::document()))
+                .change_context(DeserializerError)),
+        }
```

Comment:
Is there a way to use `try_deserialize!` here as well? I guess it's unused because `visit_null` does not take a parameter? If yes, anything we can do here?

Reply:
I don't know how to interpret the 👍🏻, sorry 😄

Reply:
I am going to try and see if I can restructure the macro a bit to be able to also be used for the Null variant^^

Reply:
Great! 
No need to put too much effort into that, but if it's possible let's do that 🚀

Reply:
fixed in [111ae6e](https://github.com/hashintel/hash/pull/1478/commits/111ae6e0af91a584e2b94e668287f1ec7034657f)

Reply:
Great!